### PR TITLE
feat(editor): Workflow history [WIP]- Add restore and clone into new workflow actions (no-changelog)

### PR DESCRIPTION
**Pull request by cstuncsik**

### CODE REVIEW

**Review by krynble — DISMISSED** [review]
Tested and it's working fine - just to make sure, testing more thoroughly is coming in the next PR right?

File: packages/editor-ui/src/stores/workflowHistory.store.ts [hunk 1]
```diff
@@ -40,9 +45,65 @@ export const useWorkflowHistoryStore = defineStore('workflowHistory', () => {
 			return null;
 		});
 
+	const downloadVersion = async (workflowId: string, workflowVersionId: WorkflowVersionId) => {
+		const [workflow, workflowVersion] = await Promise.all([
+			workflowsStore.fetchWorkflow(workflowId),
+			getWorkflowVersion(workflowId, workflowVersionId),
+		]);
+		if (workflowVersion?.nodes && workflowVersion?.connections && workflow) {
```
Comment (ivov) [inline packages/editor-ui/src/stores/workflowHistory.store.ts:53]: Why are we checking for the presence of connections? What if the workflow has nodes but no connections? Same applies below.

File: packages/editor-ui/src/stores/workflowHistory.store.ts [hunk 2]
```diff
@@ -40,9 +45,65 @@ export const useWorkflowHistoryStore = defineStore('workflowHistory', () => {
...
+			const { connections, nodes } = workflowVersion;
+			const blob = new Blob([JSON.stringify({ ...workflow, nodes, connections }, null, 2)], {
+				type: 'application/json;charset=utf-8',
+			});
+			saveAs(blob, `${workflow.name.replace(/[^a-zA-Z0-9]/gi, '_')}-${workflowVersionId}.json`);
```
Comment (ivov) [inline packages/editor-ui/src/stores/workflowHistory.store.ts:58]: Why are we so strict about filename chars? I'd expect non-ASCII chars should be allowed when saving to the filesystem? Non-English speakers are likely to use them when naming a workflow.

File: packages/editor-ui/src/stores/workflowHistory.store.ts (hunk 1 above)
Comment (ivov) [reply packages/editor-ui/src/stores/workflowHistory.store.ts:53]: On that note, what do we want to occur if the user happens to have a workflow with no nodes? I'd expect this empty workflow should be treated the same as a normal workflow.

File: packages/editor-ui/src/views/WorkflowHistory.vue [hunk 3]
```diff
@@ -22,6 +21,12 @@ type WorkflowHistoryActionRecord = {
 	[K in Uppercase<WorkflowHistoryActionTypes[number]>]: Lowercase<K>;
 };
 
+const enum WorkflowHistoryVersionRestoreModalActions {
+	restore = 'restore',
+	deactivateAndRestore = 'deactivateAndRestore',
+	cancel = 'cancel',
+}
```
Comment (ivov) [inline packages/editor-ui/src/views/WorkflowHistory.vue:28]: We should avoid raw enums.

- https://fettblog.eu/tidy-typescript-avoid-enums/
- https://www.executeprogram.com/blog/typescript-features-to-avoid
- https://dev.to/ylerjen/use-typescript-const-assertion-instead-of-enums-mfn

**Review by ivov — COMMENTED** [review]
1. Cloning does not offer a link to go to the new workflow in the toast? Or inform the name of the new workflow in the modal? I feel we are leaving the user to "go find it" rather than making it easy for them to locate it.
2. Cloning twice doesn't work? Getting `There is already an entry with this name`
3. Restoring a workflow to the latest version shouldn't work? I'd expect `The workflow is already at the latest version`
5. Restoring does not display the new topmost version? The topmost version remains the same until refreshing.
6. A workflow saved only once shows no versions instead of one version?
7. Saving an activated workflow fails to reflect the new version on the column?
8. Any answer about [these](https://n8nio.slack.com/archives/C05R4BR659R/p1696503289110479?thread_ts=1696435047.544539&cid=C05R4BR659R)?

File: packages/editor-ui/src/stores/workflowHistory.store.ts (hunk 1 above)
Comment (cstuncsik) [reply packages/editor-ui/src/stores/workflowHistory.store.ts]: Makes sense, removed the conditions

File: packages/editor-ui/src/stores/workflowHistory.store.ts (hunk 2 above)
Comment (cstuncsik) [reply packages/editor-ui/src/stores/workflowHistory.store.ts]: I'm fine with downloading the file with the workflow name as is, updated

File: packages/editor-ui/src/views/WorkflowHistory.vue (hunk 3 above)
Comment (cstuncsik) [reply packages/editor-ui/src/views/WorkflowHistory.vue:28]: Yes, I'm aware and also eslint tells you about it that's why it is `const enum` and not just `enum`

**Comment by cstuncsik** [comment]
> 1. Cloning does not offer a link to go to the new workflow in the toast? Or inform the name of the new workflow in the modal? I feel we are leaving the user to "go find it" rather than making it easy for them to locate it.
> 2. Cloning twice doesn't work? Getting `There is already an entry with this name`
> 3. Restoring a workflow to the latest version shouldn't work? I'd expect `The workflow is already at the latest version`
> 4. Restoring does not display the new topmost version? The topmost version remains the same until refreshing.
> 5. A workflow saved only once shows no versions instead of one version?
> 6. Saving an activated workflow fails to reflect the new version on the column?
> 7. Any answer about [these](https://n8nio.slack.com/archives/C05R4BR659R/p1696503289110479?thread_ts=1696435047.544539&cid=C05R4BR659R)?

These should be addressed in another PR, forwarded them to UI/UX

**Review by krynble — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.11.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.11.0)

**Pull request merged**